src/runtime_intrinsics.c

```diff
@@ -1629,8 +1629,8 @@ cvt_iintrinsic(LLVMFPtoUI, fptoui)
 #define fintrinsic_read_float32(p)   *(float *)p
 #define fintrinsic_read_float64(p)   *(double *)p
 
-#define fintrinsic_write_float16(p, x)  *(uint16_t *)p = float_to_half(x)
-#define fintrinsic_write_bfloat16(p, x) *(uint16_t *)p = float_to_bfloat(x)
+#define fintrinsic_write_float16(p, x)  *(uint16_t *)p = double_to_half(x)
+#define fintrinsic_write_bfloat16(p, x) *(uint16_t *)p = double_to_bfloat(x)
```

Comment:
I don't think the same logic applies to bfloat16. Since bfloat16 is just a truncated float32, shouldn't it always be legal to go through Float32? If not, it'd be good to have a test.

Reply:
According to the double_to_bfloat code, the same correction to the rounding direction is required whenever the float32 (after truncation from float64) is exactly halfway between 2 bfloat16 values, but wasn't required for the subnormal case (as was demonstrated in the original issue for float16)